### ISSUE TITLE
MVVM Toolkit - WPF CommandManager RelayCommand `canExecute` Issue

### ISSUE THREAD

**Issue by michael-hawker** [windows-toolkit/MVVM-Samples#41]
If you migrate a WPF application and swap the `using Galasoft.MvvmLight.CommandWpf;` to `using Microsoft.Toolkit.Mvvm.Input;` all your commands using `canExecute` will stop working due to the specific handling of `CommandManager` missing in Windows Community Toolkit MVVM framework!

_Originally posted by @laurentkempe in https://github.com/windows-toolkit/MVVM-Samples/pull/14#r523771777_

**Comment by PeskovV** [comment]
Hi everyone!
I read current discussion and understood that i should call ``NotifyCanExecuteChanged()`` in setter for correct work CanExecute method (i use RelayCommand from Microsoft.Toolkit.Mvvm)

If I'm wrong, can you correct me?

**Comment by Sergio0694** [comment]
That is correct, by default `RelayCommand` will not automatically update its visual state on WPF due to how the interface is managed there and its tight coupling with the `CommandManager` class, which we couldn't fully support in the MVVM Toolkit given that it's platform agnostic, and that type is not part of the standard BCL but it's just WPF specific.

You can either manually handle that, or use an extension to enable the auto-updating of that, like [this one I wrote](https://gist.github.com/Sergio0694/130bc344e552e501563546454bd4e62a).

The idea with that is that you'd use it to enable the auto-update for each element individually, like this:

```xml
<Button
    xmlns:input="using:Microsoft.Toolkit.Mvvm.Wpf.Input"
    Command="{Binding MyCommand}"
    input:RelayCommandExtensions.IsCommandUpdateEnabled="True"/>
```

Let me know if it works for you!

> **NOTE:** that extension type is just an example I put together, it's not official and it's not paart of the MVVM Toolkit. It's just provided as-is if you want to try it out, but if you have issues with it, please don't report that in the actual WCT repo 🙂

**Comment by PeskovV** [comment]
@Sergio0694 thank you very much!
Your extension looks like a good way for me.

I tried to use it and it has bug, i think. When control with command and extension is initializing, condition of 97 line is false (because obj doesn't have a Command)

``
/// <summary>
        /// Toggles the <see cref="IsCommandUpdateEnabled"/> logic for a target <see cref="DependencyObject"/> instance.
        /// </summary>
        /// <param name="obj">The target <see cref="DependencyObject"/> to toggle the property for.</param>
        private static void ToggleIsCommandUpdateEnabled(DependencyObject obj)
        {
            // Before doing anything, ensure previous handlers are removed
            if (handlers.TryGetValue(obj, out var data))
            {
                _ = handlers.Remove(obj);

                CommandManager.RequerySuggested -= data.CommandManagerHandler;
                data.Command.CanExecuteChanged -= data.CommandHandler;
            }

            // Register the new handlers, if requested
            if (((ICommandSource)obj).Command is IRelayCommand command &&
                GetIsCommandUpdateEnabled(obj))
            {
                object dummy = new();

                EventHandler
                    commandManagerHandler = (_, _) => CommandManager_RequerySuggested(command, dummy),
                    commandHandler = (_, _) => Command_CanExecuteChanged(dummy);

                handlers.Add(obj, new SubscriptionData(command, dummy, commandManagerHandler, commandHandler));

                CommandManager.RequerySuggested += commandManagerHandler;
                command.CanExecuteChanged += commandHandler;
            }
        }
``

**Comment by Sergio0694** [comment]
Yeah that can often happen, pretty sure I had added some code to account for that though? 🤔
I'm referring to this code here ([lines 68 to 76](https://gist.github.com/Sergio0694/130bc344e552e501563546454bd4e62a#file-relaycommandextensions-cs-L68-L76)):

```csharp
// Ensure a notifier is enabled so that we can monitor for changes to the command
// property in the future. This is also needed in case this attached property is
// set before the binding engine has actually resolved the target command instance.
if (!notifiers.TryGetValue(obj, out _))
{
    PropertyChangeNotifier notifier = new(obj, nameof(ButtonBase.Command));

    notifier.PropertyChanged += (_, _) => ToggleIsCommandUpdateEnabled(obj);
}
```

So even though you were seeing the property being `false` there, it should automatically update it when the `Command` binding is updated by the frameword. I do remember testing this and it seemed to be working fine for me. Is this actually not working at all in your application? If that's the case I could see if I can find my original sample and try it out again I guess, let me know!

**Comment by PeskovV** [comment]
You're right, but after creating command

```
DownloadCommand = new AsyncRelayCommand(DownloadCommandExecute, DownloadCommandCanExecute);

private IAsyncRelayCommand _downloadCommand;
public IAsyncRelayCommand DownloadCommand
{
    get => _downloadCommand;
    set
    {
        _downloadCommand = value;
        OnPropertyChanged(nameof(DownloadCommand));
    }
}
```
event handler ``ToggleIsCommandUpdateEnabled`` doesn't work

**Comment by Sergio0694** [comment]
Potentially unrelated, but why is your command raising a notification?
Ideally it'd just be an init-only property (or a get-only autoproperty) that's just set in the constructor. There's really no reason to make the whole property observable like that if you're never going to change its value anyway (which is generally the case) 🙂

**Comment by PeskovV** [comment]
I agree)
I test why ``ToggleIsCommandUpdateEnabled `` doesn't work and made this property observable
In generally case i use command like this:

```
SaveCommand = new RelayCommand(SaveCommandExecute, SaveCommandCanExecute); // in constructor

public IRelayCommand SaveCommand { get; private set; }
```

**Comment by Sergio0694** [comment]
You can ditch that `private set`, it's not needed in that case 🙂
Get-only autoproperty will already be settable from a constructor.

Anyway I'll see if I can repro the issue in the next few days 👍
In the meantime there's also that other variant of that class that Fons shared here: https://github.com/windows-toolkit/MVVM-Samples/issues/41#issuecomment-737070043.
You can also try using that one and see if that works as expected for you (the implementation is different so it should be fine) 😄

**Comment by PeskovV** [comment]
@Sergio0694 if i create boolean property in view model `public bool Test => true;` and then make binding `utils:RelayCommandExtensions.IsCommandUpdateEnabled="{Binding Test}"`,  i get working version

**Comment by PeskovV** [comment]
@Sergio0694 Hi! I'm sorry for the inconvenience
Have you reproduced issue? You said about it in this comment: https://github.com/windows-toolkit/MVVM-Samples/issues/41#issuecomment-808928952

**Comment by cjmurph** [comment]
@Sergio0694 
After hooking up the property changed event on the new notifier, it needs to be added to the notifiers collection or it will be collected and disposed so when the command property is updated, it won't fire ToggleIsCommandUpdateEnabled

        // Ensure a notifier is enabled so that we can monitor for changes to the command
        // property in the future. This is also needed in case this attached property is
        // set before the binding engine has actually resolved the target command instance.
        if (!notifiers.TryGetValue(d, out _))
        {
            PropertyChangeNotifier notifier = new(d, nameof(ICommandSource.Command));

            notifier.PropertyChanged += (_, _) => ToggleIsCommandUpdateEnabled(d);
            
            notifiers.Add(obj, notifier);
        }

**Comment by IngoBleile** [comment]
Will there be a fix available soon?

**Comment by Sergio0694** [comment]
@IngoBleile what are you referring to? If you mean a fix for this extension, as I mentioned that's just a gist I put together as a proof of concept, you're free to take it and modify it with the suggestions above to make it work for you. If you meant a fix to make this all work out of the box on WPF, that's not planned due to the MVVM Toolkit being cross platform by design 🙂

**Comment by IngoBleile** [comment]
@Sergio0694 Yes, I meant to make it work for WPF out of the box. Sorry, I missed the point of being cross platform! :)

**Comment by Sergio0694** [comment]
No worries! 🙂
And yeah, we can't do that because that'd require the use of `CommandManager`, which is WPF-specific unfortunately.

**Comment by will-scc** [comment]
I've come across this issue, and I understand I need to call NotifyCanExecuteChanged manually for each command in the relevant property setters, but I just want to double check I've understood fully before I commit to working on these changes.

This is how the commands are declared:

```
///Uses PropertyChanged.Fody to inject NotifyPropertyChanged
public string SomeValue { get; set; }

private ICommand addCmd;

public ICommand AddCmd => this.addCmd ?? (this.addCmd = new RelayCommand(() =>
{
    //Do some stuff
}, this.AddRecordCmd_CanExecute));

private bool AddCmd_CanExecute()
{
    return this.SomeValue == "Foo";
}
```

Since `NotifyCanExecuteChanged` isn't a part of `ICommand` interface I need to change my `ICommand` to `RelayCommand` and also convert my auto property to a full property so that I can call `NotifyCanExecuteChanged()`, like this:

```
///Uses PropertyChanged.Fody to inject NotifyPropertyChanged
private string someValue;

public string SomeValue
{ 
    get => this.someValue;
    set
    {
        this.someValue= value;
        this.AddCmd.NotifyCanExecuteChanged();
    }
}

private ICommand addCmd;

public ICommand AddCmd => this.addCmd ?? (this.addCmd = new RelayCommand(() =>
{
    //Do some stuff
}, this.AddRecordCmd_CanExecute));

private bool AddCmd_CanExecute()
{
    return this.SomeValue == "Foo";
}
```

Is this correct?

Changing from ICommand to RelayCommand is easy enough, it's having to update all the properties used in `CanExecute` methods that's unpleasant with how many commands there are and each one having to be done by hand (rather than a nice easy Replace job).

Side note: it's disappointing that the [Migrating from Mvvmlight](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/migratingfrommvvmlight) article doesn't mention this breaking change at all. :(

**Comment by pikausp** [comment]
> No worries! 🙂 And yeah, we can't do that because that'd require the use of `CommandManager`, which is WPF-specific unfortunately.

Hello Sergio, you guys are obviously free to handle things how you see fit, but would you say MAUI, for example, is not cross platform due to it allowing you to tweak platform specific things? Personally I would say that having support for command re-query for WPF is something that many developers have use for and it takes away from the added benefit of the library. If we were talking about some obscure feature I would understand, but this seems as sort of a core functionality on one of the platforms.

Just a food for thought :)

**Comment by doxxx** [comment]
Another alternative solution is to call `NotifyCanExecuteChanged` on `IRelayCommand`s when `CommandManager.RequerySuggested` is raised:

```cs
CommandManager.RequerySuggested += (s, e) =>
{
    foreach (var command in GetType().GetProperties()
        .Where(p => p.PropertyType.IsAssignableTo(typeof(IRelayCommand)))
        .Select(p => (IRelayCommand?)p.GetValue(this)))
    {
        command?.NotifyCanExecuteChanged();
    }
};
```

You'd have to do this on each type that exposes `IRelayCommand` properties, but you could wrap it up as a utility method that you just call from the constructor after the properties are initialized.

**Comment by doxxx** [comment]
I created a gist for this helper: https://gist.github.com/doxxx/911ba2af9847362e12e01a045273184e
